### PR TITLE
WikiText-103 and undocumened MultNLI

### PR DESCRIPTION
**Pull request by PiotrCzapla**
- The wikitext-103 is being tested on sotabench at the moment
https://sotabench.com/user/piotr.czapla-priv#builds

Docs have an additional TLDR section that shows a minimal example.

I've added aproposal to include "eval" function that encapsulate the add > cache > save logic

### CODE REVIEW

File: docs/docs/wikitext103.md [hunk 1]
```diff
@@ -0,0 +1,205 @@
+# WikiText-103
+
+![Banner](img/banner.png)
+[//]: # (TODO Change Me)
```
Comment (RJT1990) [inline docs/docs/wikitext103.md]: please fix - you can mirror the style of the other task docs closely (e.g. imagenet and coco)

File: docs/docs/wikitext103.md [hunk 2]
```diff
@@ -0,0 +1,205 @@
+# WikiText-103
+
+![Banner](img/banner.png)
+[//]: # (TODO Change Me) 
+
+You can download WikiText-103 dataset [here](https://blog.einstein.ai/the-wikitext-long-term-dependency-language-modeling-dataset/), there is no official leaderboard.
```
Comment (RJT1990) [inline docs/docs/wikitext103.md]: link to the sotabench leaderboard (as with the other docs for each task)

File: docs/docs/wikitext103.md [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+
+## TLDR;
```
Comment (RJT1990) [inline docs/docs/wikitext103.md]: ? please use same format as other (imagenet and coco) for consistency. I think we use the heading "a full sotabench.py example" (or something like that) and it's the last section

File: docs/docs/wikitext103.md [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+Here is a minimal working example that let you evaluate `transformer-xl` from `huggingface/transformers`.
+
+```python
+import torch
+from sotabencheval.langauge_modelling import WikiText103Evaluator
+
+model = torch.hub.load('huggingface/transformers', 'modelWithLMHead', 'transfo-xl-wt103').to("cuda")
+tokenizer = torch.hub.load('huggingface/transformers', 'tokenizer', 'transfo-xl-wt103').to("cuda")
+
+with WikiText103Evaluator.dataset.get_path('local/dir/wikitext-103').open() as f:
```
Comment (RJT1990) [inline docs/docs/wikitext103.md]: please use `is_server()` so it's not just using local paths (see coco/imagenet docs). Also the use of a context manager here is a bit confusing, this needs to be documented (and if API can be simpler that's great).

File: docs/docs/wikitext103.md [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+        X, Y, mems = test_data[None, :-1], test_data[None, 1:], None
+        for s in range(0, X.shape[-1], seq_len):
+            log_probs, mems, *_ = model(input_ids=X[s:s+seq_len], mems=mems)
+            yield log_probs, y[s:s+seq_len]
+
+WikiText103Evaluator(
+    model_name="Transformer-XL Large", 
+    paper_arxiv_id="1901.02860",
+    paper_pwc_id="transformer-xl-attentive-language-models" 
+).eval(log_probs_generator(model, test_data)).print_results()
```
Comment (RJT1990) [inline docs/docs/wikitext103.md]: confused. also there's no evaluator.save? Ideally we would want consistent API between different benchmarks.

File: docs/docs/wikitext103.md [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+
+Your local files may have a different file directory structure, so you
+can use control flow like below to change the data path if the script is being
+run on sotabench servers:
+
+``` python
+from sotabencheval.utils import is_server
+
+if is_server():
+    DATA_ROOT = '.data/nlp/squad'
```
Comment (RJT1990) [inline docs/docs/wikitext103.md]: wikitext-103?

File: docs/docs/wikitext103.md [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+else: # local settings
+    DATA_ROOT = '/home/ubuntu/my_data/'
+```
+This will detect if `sotabench.py` is being run on the server and change behaviour accordingly.
+
+Alternatively you can use our helper function that will automatically unpack the zip for you and point you to the right location as follows:
+
+```python
+from sotabencheval.langauge_modelling import WikiText103Evaluator
```
Comment (RJT1990) [inline docs/docs/wikitext103.md]: typo

File: docs/docs/wikitext103.md [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+
+test_file_path = WikiText103Evaluator.dataset.get_path('/home/ubuntu/data')
+```
```
Comment (RJT1990) [inline docs/docs/wikitext103.md:60]: Unzipping automatically is fine, but not sure about this get_path method. in any case, can we pass DATA_ROOT into this? also it's confusing because I don't know what I'm supposed to do with this?

File: docs/docs/wikitext103.md [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+
+## How Do I Initialize an Evaluator?
+
+Add this to your code - before you start batching over the dataset and making predictions:
+
+``` python
+from sotabencheval.langauge_modelling import WikiText103Evaluator
```
Comment (RJT1990) [inline docs/docs/wikitext103.md]: typo

File: docs/docs/wikitext103.md [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+evaluator = WikiText103Evaluator(model_name='Model name as found in paperswithcode website')
+```
+
+If you are reproducing a model from a paper, then you can enter the arXiv ID. If you
+put in the same model name string as on the
+[Wikitext-103](https://sotabench.com/benchmarks/language-modelling-on-wikitext-103) leaderboard
+then you will enable direct comparison with the paper's model. If the `arxiv` is not avaliable you can use `paperswithcode.com` id.  Below is an example of an evaluator that matches `Transformer XL`:
+
+``` python
+from sotabencheval.langauge_modelling import WikiText103Evaluator
```
Comment (RJT1990) [inline docs/docs/wikitext103.md]: typo

File: docs/docs/wikitext103.md [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+    paper_arxiv_id="1901.02860",
+    paper_pwc_id="transformer-xl-attentive-language-models"
+)
+```
+
+The above will directly compare with the result of the paper when run on the server.
+
+## How Do I Evaluate Predictions?
+
+The evaluator object has an `.add(log_probabilites, labels)` method to submit predictions by batch or in full. The method except two `np.arrays` or `torch.tensors` of the same shape `[batch_size, sequence_length]`, first one holding `log` probabilities of each token and second holding token ids.
```
Comment (RJT1990) [inline docs/docs/wikitext103.md]: I'm confused because here it is add, but before you were using eval?

File: docs/docs/wikitext103.md [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+
+The above will directly compare with the result of the paper when run on the server.
+
+## How Do I Evaluate Predictions?
+
+The evaluator object has an `.add(log_probabilites, labels)` method to submit predictions by batch or in full. The method except two `np.arrays` or `torch.tensors` of the same shape `[batch_size, sequence_length]`, first one holding `log` probabilities of each token and second holding token ids. 
+
+If your model use subword tokenization you are free to use them without the need to convert them to full words, but make sure to mark that in your evaluator with `subword_tokenization=True`, as we are including some checks to make the perplexity calculation less error prone.
+
+[//]: # (TODO Describe how to ensure that the right tokenization is being used)
```
Comment (RJT1990) [inline docs/docs/wikitext103.md]: ?

File: sotabencheval/core/evaluator.py [hunk 13]
```diff
@@ -78,14 +78,39 @@ def cache_exists(self):
...
 
+    def reset(self):
+        """Resets the internal state of evaluator and allows to start over"""
+        pass
+
     def cache_values(self, **kwargs):
+        """Convert kwargs to something that can be used to generate batch hash"""
         return cache_value(kwargs)
 
+    def eval(self, results_generator):
```
Comment (RJT1990) [inline sotabencheval/core/evaluator.py:170]: I have no idea what the results_generator should look like - if we are to do things this way, then it needs to be documented both in doc strings, and also very clearly in the docs themselves

File: sotabencheval/language_modelling/wikitext.py [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+        if isinstance(a, list):
+            return np.array(a)
+        return a 
+    return [convert(a) for a in args]
+
+class WikiTextDataset(Enum):
+    WikiText103 = ('WikiText-103', 245569, 267735)
+    WikiText2 = ('WikiText-2', 245569, 33278)
+    
+    def __init__(self, pwc_name, testset_size, vocab_size):
```
Comment (RJT1990) [inline sotabencheval/language_modelling/wikitext.py:20]: docstrings

File: sotabencheval/language_modelling/wikitext.py [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+        self.pwc_name = pwc_name
+        self.testset_size = testset_size
+        self.vocab_size = vocab_size
+    
+    def get_path(self, local_root, local_unzip=False):
+        root = Path(change_root_if_server(root=local_root,
```
Comment (RJT1990) [inline sotabencheval/language_modelling/wikitext.py:32]: what does this method do? where is it used?  docstrings

File: sotabencheval/language_modelling/wikitext.py [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+                                          server_root=".data/nlp/" + self.pwc_name.lower()))
+        zip_name = self.pwc_name.lower() + "-v1.zip"
+        dataset_path = root / "wiki.test.tokens"
+        if not dataset_path.exists(): # unzip
+            extract_archive(str(root / zip_name), to_path=root.parent)
+        return dataset_path
+
+
+def gether_probs(log_probabilities, targets):
```
Comment (RJT1990) [inline sotabencheval/language_modelling/wikitext.py]: ?

File: sotabencheval/language_modelling/wikitext.py [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+                 dataset=None):
+
+        super().__init__(model_name, paper_arxiv_id,
+                         paper_pwc_id, paper_results, model_description)
+        if dataset is not None:
+            self.dataset = dataset
+        self.subword_tokenization = subword_tokenization
+        self.text_transformation = text_transformation
+        self.local_root = local_root 
+        self.reset()
```
Comment (RJT1990) [inline sotabencheval/language_modelling/wikitext.py]: why reset?

File: sotabencheval/language_modelling/wikitext.py [hunk 18]
```diff
@@ -0,0 +1,132 @@
...
+        
+    def reset(self):
+        self._neglogloss = 0
+        self._data_set_size = 0 
+    
+    def add(self, log_probabilities, targets):
+        """
+            log_probabilietes - 
+                float - summed log probability of targets
+                [bs x bptt] array of floats - log probability of each target log_proabilites.shape == targets.shape
```
Comment (RJT1990) [inline sotabencheval/language_modelling/wikitext.py]: np.ndarray?

Comment (RJT1990) [reply sotabencheval/language_modelling/wikitext.py]: can we have an example of what the input looks like?

**Review by RJT1990 — COMMENTED** [review]
Hey Piotr - thanks for the PR.

Comments:

- Unless absolutely necessary, I would not try to deviate too far from the API style of the other benchmarks - e.g. using new methods like eval, not using .save() etc. We might want to change these in future but it adds more complexity to have a completely independent approach to doing benchmarks

- If it is necessary to deviate because the evaluation procedure for language modelling is so different, then it needs to be crystal clear in the documentation what is happening - both in the docs themselves, and docstrings. This might require a step-by-step example; it also requires more context on why and where different methods are used, and may also require some example inputs and outputs for different methods.

- In terms of missing things, ideally we would want the speed calculation on the platform before the launch (so we can see speed vs perplexity of the different models).

File: docs/docs/wikitext103.md (hunk 2 above)
Comment (PiotrCzapla) [reply docs/docs/wikitext103.md]: >Unless absolutely necessary, I would not try to deviate too far from the API style of the other benchmarks - e.g. using new methods like eval, not using .save() etc. We might want to change these in future but it adds more complexity to have a completely independent approach to doing benchmarks

Ross that is just a proposal, both  API works, the eval is added to the base  class so it is available everywhere where the base class is used. I can remove that from the docs and I can remove the code. I understand you don't like the approach with  generators, i suppose you think it makes it more complicated for arbitrary ml person to follow, although this approach come with prons as well:
- you can't forget about  `save` or `cache_exists`
- it gives us more control over the calls to the model, it should be easier to measure the memory usage before first batch.
- it will be easier to catch other statistics, and to catch errors during execution.

But as I said above it is just a proposal I wanted you to see how it would look like, you find it complex let's have it removed.

>  speed calculation on the platform before the launch (so we can see speed vs perplexity of the different models).
I though it was added to the base class, good catch let me make sure it is working.

File: docs/docs/wikitext103.md (hunk 3 above)
Comment (PiotrCzapla) [reply docs/docs/wikitext103.md]: Good catch, was doing that late and I though by leaderboards you are referring to the external leader boards, hence the adopted text. i will change that to our leader boards right away.

File: docs/docs/wikitext103.md (hunk 4 above)
Comment (PiotrCzapla) [reply docs/docs/wikitext103.md]: The `get_path` method has the following signature `(local_path:StrOrPath) -> Path`. The `is_server` is being used in the `get_path` method, and the path to the `wikitext-103-v1.zip`  is hardcoded in the method, along with the use of `extract_archive` that unpack the zip if needed.

The context manager and `.open` is the idiom from `pathlib.Path`. 
We have the following ways already implemented to get path to the dataset all of them are taking care of extracting the archive let me know which are better:

If you want to read the dataset before you instantiate the evaluator you can do this on 3 ways:
Using is_server and rewrting the logic of get_path:

```python
if is_server():
      root = Path(".data/nlp/wikitext-103")
else:
      root = Path("/path/to/your/location/where/you/have/unzipped/wikitext103")
dataset_path = root / "wiki.test.tokens"
if not dataset_path.exists(): 
       # on server we need to unzip the wikitext first.
       # note the use of the root.parent as the .zip file has already wikitext-103 folder in it
       # without this root.parent you need to set the path to `.data/nlp/wikitext-103/wikitext-103/wiki.test.tokens `
       extract_archive(str(root / "wikitext-103-v1.zip"), to_path=root.parent)
with dataset_path.open() as f:
    testset_content = f.read()
```

Or we can simplify the code using the convenience method of `dataset.get_path` 

```python
path_to_testset =  WikiText103Evaluator.dataset.get_path(local_root="/path/to/your/location/where/you/have/unzipped/wikitext103")
with path.open() as f:
    testset_content = f.read()
```

The dataset is actually an instance of WikiTextDataset enum so we could use that directly:

```python
path_to_testset = WikiTextDataset.Wikitext103.get_path(local_root="/path/to/your/location/where/you/have/unzipped/wikitext103")
with path.open() as f:
    testset_content = f.read()
```

Things might be even simpler if we read the dataset after the WIkiText103Evaluator was created:


```python
evaluator= WikiText103Evaluator(local_root="/path/to/your/local/data", model_name ...)
with  evaluator.dataset_path.open() as f:
      testset_content = f.read()
```



Given that we may want to evaluate on Validation set as well we may want to return paths and change the `dataset_path` property to `test_set_path` or `validation_set_path`

And change the `WikiTextDataset.WikiText103.get_path` to `get_test_set_path` and `get_validation_set_path`.

I'll ask on slack which api is better and the right example in to the docs.

File: docs/docs/wikitext103.md (hunk 5 above)
Comment (PiotrCzapla) [reply docs/docs/wikitext103.md]: This just an example of a wrapper that could simplify things as you don't have to remember about `save`, `check_cache`, `reset`, `reset_timer` and exception handling.

Comment (PiotrCzapla) [reply docs/docs/wikitext103.md]: But as said before, I will update docs to have the same api.

File: sotabencheval/language_modelling/wikitext.py (hunk 17 above)
Comment (PiotrCzapla) [reply sotabencheval/language_modelling/wikitext.py]: It initialise the counters i can copy the code here.

File: sotabencheval/language_modelling/wikitext.py (hunk 15 above)
Comment (PiotrCzapla) [reply sotabencheval/language_modelling/wikitext.py:32]: it is the methods that convience method to extract the dataset `WikText103Evaluator.dataset.get_path`. Will add the docs strings.

File: sotabencheval/language_modelling/wikitext.py (hunk 16 above)
Comment (PiotrCzapla) [reply sotabencheval/language_modelling/wikitext.py]: Depending on the level of elasticity in the model code one can use 3 levels of validation(the 
 validation checks are yet to be written though).
- level 1: you just provide a summed log probability over a given batch of sentences (batch of sentences are in targets)
- level 2: if your model lets you get a probability for each word /token in the batch that is even better we accept that.
- level 3:  but ideally we would like to see your prediction of log probabilities of all possible words for each word in a batch. os we allow the input to be of shape: `[batch_size x seq_length x vocab_size]`. There we need the gather method implemented in `np` and `torch` (For speed).

File: docs/docs/wikitext103.md (hunk 2 above)
Comment (RJT1990) [reply docs/docs/wikitext103.md]: Sure - no problem. I understand the arguments. I think I would opt for between-benchmark-consistency in this case, just because it's additional overhead if there's two ways of doing things (even if one takes care of more things automatically!). We can discuss how we can improve the API in general from next week onwards; might be that we adopt something closer to this proposal!

File: docs/docs/wikitext103.md (hunk 5 above)
Comment (RJT1990) [reply docs/docs/wikitext103.md]: yep makes sense, but let's keep things consistent for now and improve things later. One way to think about this is that we'll get feedback on the API upon release and how people are using it. Then we can figure out what to keep, and what we don't need. Then once we have this high quality feedback, we can figure out what to wrap and simplify (similar to your suggestions).

File: docs/docs/wikitext103.md (hunk 4 above)
Comment (RJT1990) [reply docs/docs/wikitext103.md]: Yep I think something like this makes sense. Makes sense that you need something like this for language modelling where the user needs more control over the data - e.g. needs to tokenise the corpus with their own logic.

Just needs to be documented clearly - I think a step-by-step example is the best way to explain why we need this.

File: sotabencheval/language_modelling/wikitext.py (hunk 15 above)
Comment (RJT1990) [reply sotabencheval/language_modelling/wikitext.py:32]: ty - and also a step-by-step example where this is used

File: sotabencheval/language_modelling/wikitext.py (hunk 16 above)
Comment (RJT1990) [reply sotabencheval/language_modelling/wikitext.py]: Okay, just needs to be documented (docstrings, and example usage in the docs themselves).

File: sotabencheval/language_modelling/wikitext.py (hunk 17 above)
Comment (RJT1990) [reply sotabencheval/language_modelling/wikitext.py]: might be better to copy the code just to be explicit
